mypy/stubgen.py

```diff
@@ -222,25 +226,141 @@ def generate_stub(path: str, output_dir: str, _all_: Optional[List[str]] = None,
+        if (not self.requires_quotes and
+                base not in dir(builtins) and
+                not any(base in vs for vs in self.stubgen._vars)):
+            self.requires_quotes = True
+
+        if t.args != []:
+            s += '[{}]'.format(self.list_str(t.args))
+        return s
+
+    def visit_none_type(self, t: NoneTyp) -> str:
+        return "None"
+
+
+class AliasPrinter(NodeVisitor[str]):
+
+    def __init__(self, stubgen: 'StubGenerator') -> None:
+        self.stubgen = stubgen
+        super().__init__()
+
+    def visit_call_expr(self, node: CallExpr) -> str:
+        callee = node.callee.accept(self)
+        args = []
+        for name, arg, kind in zip(node.arg_names, node.args, node.arg_kinds):
+            if kind == ARG_POS:
+                args.append(arg.accept(self))
+            elif kind == ARG_STAR:
+                args.append('*' + arg.accept(self))
+            elif kind == ARG_STAR2:
+                args.append('**' + arg.accept(self))
+            elif kind == ARG_NAMED:
+                args.append('{}={}'.format(name, arg.accept(self)))
+            else:
+                raise ValueError("Unknown argument kind %d in call" % kind)
+        return "{}({})".format(callee, ", ".join(args))
+
+    def visit_name_expr(self, node: NameExpr) -> str:
+        self.stubgen.import_tracker.require_name(node.name)
+        return node.name
+
+    def visit_str_expr(self, node: StrExpr) -> str:
+        return repr(node.value)
+
+    def visit_index_expr(self, node: IndexExpr) -> str:
+        base = node.base.accept(self)
+        index = node.index.accept(self)
+        return "{}[{}]".format(base, index)
+
+    def visit_tuple_expr(self, node: TupleExpr) -> str:
+        return ", ".join(n.accept(self) for n in node.items)
+
+    def visit_list_expr(self, node: ListExpr) -> str:
+        return "[{}]".format(", ".join(n.accept(self) for n in node.items))
+
+    def visit_ellipsis(self, node: EllipsisExpr) -> str:
+        return "..."
+
+
+class ImportTracker:
+
+    def __init__(self) -> None:
+        self.module_for = {}  # type: Dict[str, Optional[str]]
+        self.direct_imports = {}  # type: Dict[str, str]
+        self.reverse_alias = {}  # type: Dict[str, str]
+        self.required_names = set()  # type: Set[str]
+
+    def add_import_from(self, module: str, names: List[Tuple[str, Optional[str]]]) -> None:
+        for name, alias in names:
+            self.module_for[alias or name] = module
+            if alias:
+                self.reverse_alias[alias] = name
+
+    def add_import(self, module: str, alias: str=None) -> None:
+        name = module.split('.')[0]
+        self.module_for[alias or name] = None
+        self.direct_imports[name] = module
+        if alias:
+            self.reverse_alias[alias] = name
+
+    def require_name(self, name: str) -> None:
+        self.required_names.add(name.split('.')[0])
+
+    def import_lines(self) -> List[str]:
```

Comment:
It is difficult to follow the logic of this method. Maybe a doctring (and comments) will help?

Reply:
Done